### ISSUE TITLE
m3e有什么特别的改进吗？

### ISSUE THREAD

**Issue by Nipi64310** [wangyuxinwhy/uniem#22]
### 🐛 bug 说明

Hi @wangyuxinwhy ，
感谢分享，主要是有个问题想讨论一下。虽然m3e在几个开源测试集上指标有提升，但是相比openai-ada-002，抗干扰能力明显变弱

比如在自己构造的一份检索的测试集上，m3e-base几乎不能正确召回，luotuo为开源的在openai向量上蒸馏出来的向量模型，虽然效果远不如openai的结果，但是对比m3e也是优势比较大
![image](https://github.com/wangyuxinwhy/uniem/assets/31469418/079da7c5-41cb-4be2-9b02-6636e145fb21)

luotuo用你们的评测代码，效果是远不如m3e的
![image](https://github.com/wangyuxinwhy/uniem/assets/31469418/b68a4cac-a69a-43a7-a886-ca552c76909c)

测试代码如下：
每个docs是3条数据，docs[0]为问题q,  docs[1]为m3e检索回来的相似度最高的doc片段,  docs[2]为正确的doc片段
```python
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
model_path = 'moka-ai/m3e-base'
model = SentenceTransformer(model_path)

docs = ["富森美公司的主要收入模式是什么",
"""公司多年来深耕装饰建材
家居业务，合理设置卖场定位，拥有大量优质代理商、经销商客户资源， 同时，公司在集
合式营销、市场管理、资源整合、产品促销等方面不断创新， 相继被中宣部等八部委联合
授予“全国百城万店无假货示范市场” ，被 国家工商总局授予“全国诚信示范市场” ，被
国家质检总局授予 “全国购物放心市场” ，被 四川省人民政府授予“四川省服务名牌企业”
等荣誉称号， 逐渐树立起品牌规范性专业市场形象 。“富森美”品牌已成为成都及四川地
区具有广泛影响力的装饰建材流通类品牌，在消费者心��中享有较高的品牌认知度和美誉
度,有力地支持了公司渠道扩张和商业模式的实行。  


“服务 +产品”策略， 全业态定位满足消费者需求 。在区域专业市场中， 公司通过 合理的
定位、 高度的便利性和全面的 “服务 +产品” 策略 有效地吸引商户入驻和消费者体验、 消费。""",
 
"""公司目前
自持卖场物业建筑面积超过 110 万平方米， 2019 年底加盟及委托管理卖场经营规模达到
51.6 万平方米， 公司自有和委管卖场均使用的 “富森美” 品牌， 近三年获评 2018 年和 2019
年的中国家居品牌大会“中国十大家居连锁卖场”以及“ 2019 -2020 十大优选中国家居消
费平台” 。公司当前主要收入模式来自卖场招商收租金， 毛利率、 净利率和成本费用率均处
于行业前列，富森美主要优点在于： 1）自持物业成本低； 2）精细化运营能力强，布局多
层次业态，差异化招商、服务； 3）产业链整合创新能力突出，平台资源丰富。 
 数字化整合升 级， 加速新零售布局。公司紧跟家居行业个性化等需求增长的趋势， 围绕 “引
流和赋能” ，通过微信小程序、网上直播、联合营销、专场活动以及优化品牌结构等多种方
式巩固公司的区域优势，重点对卖场数字化升级，实现线上流量的精准转化，同时提升运
营效率。"""]
embeddings = model.encode(docs)
print(cosine_similarity(embeddings))


docs = ["富森美公司的风险提示有哪些",
"""公司报告  
请务必仔细阅读正文之后的各项信息披露与声明  第21页 共28页                        简单金融  成就梦想  
 
2019 年富森美在 中国上市公司口碑榜 当中荣获 “最具成长性新零售产业上市公司” ，
公司发展新零售模式主要优势在于： 1）创始人 多年从事装饰建材家居行业工作， 始终以
用户思维、产品思维而非商业地产思维发展业务 ，创新打造或孵化出多层次新业态 ；2）
品牌与销售方面， 公司 始终深耕产品，协助商户完成各类产品的市场营销活动， 为厂家和
商户提供了全产业链服务， 具备良好的销售经验和基础； 3）公司的供应链体系完整，资
源丰富，平台优势突出； 4）新零售项目运营团队年轻化，由 90后担纲运营，能够准确地
把握当代消费者的需求 。""",
 
"""我们对
公司 2020~2022 年EPS 预测分别为 0.89 元/1.09 元/1.26 元，对应 PE分别为
18X/15X/13X ，首次覆盖，给予“买入”评级。  
 风险提示： 地产景气周期下滑，市场竞争加剧，公司新零售业务发展进度不及预期。  
市场数据：  2020 年12月08日 
收盘价（元 ）  16.15  
一年内最高 /最低（元）  16.85 /10.47  
市净率  2.3 
息率（分红 /股价）  2.17 
流通 A股 市 值 （ 百 万 元 ）  4819  
上证指数 /深证成指  3410.18 /13973.89  
注：“息率 ” 以 最 近 一 年 已 公 布 分 红 计 算  
  
基础数据 : 2020 年09月30日 
每股净资产 （ 元 ）  7.02 
资产负债率 % 14."""]
embeddings = model.encode(docs)
print(cosine_similarity(embeddings))

"""
[[1.         0.8282594  0.79422724]
 [0.8282594  1.         0.89455336]
 [0.79422724 0.89455336 0.9999997 ]]

[[1.        0.8106332 0.6698204]
 [0.8106332 1.        0.8019115]
 [0.6698204 0.8019115 0.9999999]]
"""
```

### Python Version

None

**Comment by wangyuxinwhy** [comment]
我测试了你的 case ，和你的分数一致，m3e 确实在你的场景中确实表现的不好，还有另一位同学也是金融研报的场景，也是和你同样的问题，m3e 也是效果比较差。

其原因是由于 m3e 可能在学习中找到一种 “偷懒” 的方式，m3e 会过度关注 “主题” 词汇，在你的这个场景中是 **富森美** 。因为这个词不是很经常出现，所以这个 m3e 只要见到两个文本中都有这个词，就会认为两个文本是相似的，就不会进一步区分两个文本之间的细节，比如提没提到收入，风险提示等等

佐证是如下的两个例子：
![image](https://github.com/wangyuxinwhy/uniem/assets/22192665/4d260be6-6d53-4ef7-856a-91af3d77761d)
这个例子中，我们将富森美用代词代替，m3e 就可以正确的区分文本中收入的细节了
![image](https://github.com/wangyuxinwhy/uniem/assets/22192665/e251ce8f-e2b5-422a-8eaa-925b7aba5c53)
这个例子中，我们将文本中唯一出现过的 富森美 改为用代词来代替，我们可以发现相似度从 0.82 -> 0.69 

为什么会出现这种情况呢？
m3e 使用的对比学习是在一个批次内进行负采样，但这种方法采样出来的负例，很少出现同一主题但不同细节，往往是不同主题。也就是说这些负例不需要 m3e 区分细节，只需要区分主题就可以了。这就导致了 m3e 会采用 “偷懒” 的方式去判断一个文本中的主题，而忽略细节。这种判断主题的能力，也使得 m3e 在 T2Ranking 这种具有广泛主题的任务上有更好的表现。

如何解决？
归根结底是因为在构造数据的时候，没有构造 “同一主题，不同细节” 的困难负例（hard negative sample）导致的。这也恰恰是 m3e 下一次迭代的重点。

**Comment by Nipi64310** [comment]
Hi @wangyuxinwhy ，
感谢回复，测试了下，即使去掉问题中的公司名字，也处理不了我这份测试数据，
比如上例中：
```python
query:公司的风险提示有哪些

"""
公司报告  
请务必仔细阅读正文之后的各项信息披露与声明  第5页 共28页                        简单金融  成就梦想  
 
图30：数字化升级 ................................ ................................ .........................  20 
图31：新零售体系构建  ................................ ................................ ..................  20 
图32：富森美启动首个家居网红直播基地  ................................ .......................  21 
图33：打造 3万平方米体验中心  ................................ ................................ ....
"""
得到的cosine值会比正确文档片段高。

```
按照现在这个模式训练，训练一个好的embedding，负样本采样确实是一个比较大的问题。



另外，还有2个问题，
1. m3e在mteb-zh效果好，主要是因为训练数据量（2000w+pair）大吗，有其他的改进吗？我测试了一下text2vec针对检索的模型也是比m3e差挺多的。
2. 基础模型为啥选hfl的roberta呢，用其他更好的基础模型有评测过差异吗

**Comment by wangyuxinwhy** [comment]
**去掉公司也不对**
去掉公司名称也还是不对，这个有点奇怪，不知道是什么原因导致的，需要分析一下，有结论我再和你说。

**m3e在mteb-zh效果好**
我在构建 mteb-zh 的时候是在 huggingface 上面进行挑选的，原则就是 m3e 没有在训练中见过，text2vec 也没有在训练中见过。至于效果好，纯粹是因为训练的数据量比较大，别的优化是没有的。可以参考 [训练脚本](https://github.com/wangyuxinwhy/uniem/blob/main/scripts/train_m3e.py) 
另外，可能 s2p 的数据占比比较大， 所以泛主题检索场景下 m3e 是有很大优势的。

**为啥选 roberta **
之前打比赛测评过很多中文预训练模型，hfl/roberta 的表现比较稳定，所以做了这个很中庸的选择。其他的模型没有仔细的测试过，有推荐的吗？

我目前对预训练模型有两个要求，一个是能支持中英，另一个是支持长上下文（4k）。但我目力所及的开源模型没有满足这两个需求的，所以我正在自己训练一个，架构类似于 [mosaicbert](https://www.mosaicml.com/blog/mosaicbert)。

**Comment by Nipi64310** [comment]
请问下，这个脚本里面加载的数据集https://github.com/wangyuxinwhy/uniem/blob/main/scripts/process_zh_datasets.py
比https://huggingface.co/moka-ai/m3e-base readme中提到的少？
比如这2个

> webtext2019zh	百科	4,258,310	问答	问答	优	Bright Xu	含有410万个预先过滤过的、高质量问题和回复。每个问题属于一个【话题】，总共有2.8万个各式话题，话题包罗万象。	是	是	否	是	https://github.com/brightmart/nlp_chinese_corpus	否
>SimCLUE	百科	775,593	平行语义	相似	良	数据集合，请在 simCLUE 中查看	整合了中文领域绝大多数可用的开源的语义相似度和自然语言推理的数据集，并重新做了数据拆分和整理。	是	否	否	是	https://github.com/CLUEbenchmark/SimCLUE	是

或者有提供全量数据集下载的地方吗？

**Comment by wangyuxinwhy** [comment]
非 huggingface 的数据集因为需要在网站上注册，并同意一些条款，所以没有统一的下载链接。

处理脚本可以参考我脚本中的的处理方式，非常简单，一般来讲就是把字段进行重命名。

**Comment by Nipi64310** [comment]
Hi， @wangyuxinwhy 

repo里面有写了多种损失，根据train_m3e.py中最后使用的是
`self.criterion = PairInBatchNegSoftmaxContrastLoss(temperature)`，请问有对比不同损失评测效果的差异吗？

**Comment by wangyuxinwhy** [comment]
目前 Uniem 实现了三种损失函数
1. 经典对比学习损失函数 
$$
\ell_i=-\log \frac{e^{\operatorname{sim}\left(\mathbf{h}_i, \mathbf{h}_i^{+}\right) / \tau}}{\sum_{j=1}^N e^{\operatorname{sim}\left(\mathbf{h}_i, \mathbf{h}_j^{+}\right) / \tau}}
$$

**Comment by NLPJCL** [comment]
> 目前 Uniem 实现了三种损失函数
> 
> 1. `SoftmaxContrastLoss`，论文参考[simcse](https://arxiv.org/pdf/2104.08821.pdf)损失函数
>    ![](https://camo.githubusercontent.com/94152511f5919feff8e1d15845a54d4646da4cbfa80c49571aa15645fb7573a4/68747470733a2f2f797578696e2d77616e672e6f73732d636e2d6265696a696e672e616c6979756e63732e636f6d2f755069632f5a70385567612e706e67)
> 2. `SigmoidContrastLoss`，论文参考 [instrcut gpt](https://arxiv.org/pdf/2203.02155.pdf) 中提出的排序函数
>    ![](https://camo.githubusercontent.com/403bbc9756a3e5c8d0b817d2c14c2e6796cd816f639ffa6cc07a8169b2a3f9de/68747470733a2f2f797578696e2d77616e672e6f73732d636e2d6265696a696e672e616c6979756e63732e636f6d2f755069632f475377766e782e706e67)
> 3. `CoSentLoss`, 论文参考 [cosent loss](https://kexue.fm/archives/8847)
>    ![](https://camo.githubusercontent.com/e82f74739cbcfd798a294b75d917b3a58386968f4897c1232fcb2c5893cf2959/68747470733a2f2f797578696e2d77616e672e6f73732d636e2d6265696a696e672e616c6979756e63732e636f6d2f755069632f6971587437752e706e67)
> 
> 其中，不同的 Loss 有不同的使用场景，一般来讲句对或者句子三元组来说会更经常使用 `SoftmaxContrastLoss` 和 `SigmoidContrastLoss`，batch size 小的时候使用 `SigmoidContrastLoss` 会好一点，大的时候 `SoftmaxContrastLoss` 好一点。如果有明确的相似度分数，则使用 CoSent Loss 会好一点。

您好，SigmoidContrastLoss这个loss不太理解，您能给点资料或者讲讲嘛？看代码好像是把query和正例就不计算loss(对应位置mask为1e9，sigmod后就是1，-log(sigmod)后就没有loss了)，而是计算query和负例的loss？是这样嘛？
